controllers/orderController.js

```diff
@@ -17,13 +17,15 @@ const createOrder = async (req, res) => {
   }
   if (!tax || !shippingFee) {
     throw new CustomError.BadRequestError(
-      'Please provide tax and shipping fee'
+      'Please provide tax and shipping fee',
     );
   }
 
   let orderItems = [];
   let subtotal = 0;
   //if there are items in cartItems, we set up a loop
+
+  // AKOSREVIEW: Show examples of other ways to do this with *.map, *.reduce, or .forEach
   for (const item of cartItems) {
```

Comment:
This is totally fine as is! But I just wanted to also share some other ways of writing this, using the Array iterator method, [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).

```javascript
// Array.prototype.reduce takes two values, the first is a function, and the second is an "initial value"
// the function should have two arguments, the first argument is the element in the array that we're iterating over (in this case, it'll be each  order item)
// the second argument is the "accumulator" that you want to add your results to each iteration. In this case, I'm using a map to hold my  subtotal and orderItems variables that I want to modify each iteration

const initialValue = { subtotal: 0, orderItems: [] }

const { subtotal: subtotal, orderItems: orderItems } = cartItems.map((item, resultsMap => {
    const dbProduct = await Product.findOne({ _id: item.product });
    
    if (!dbProduct) {
        throw new CustomError.NotFoundError(`No product with id ${item.product}`);
    }
    
    const { name, price, image, _id } = dbProduct;
    const singleOrderItem = {
      amount: item.amount,
      name,
      price,
      image,
      product: _id,
    };
    
    resultsMap.orderItems = [...resultsMap.orderItems, singleOrderItem]
    resultsMap.subtotal += item.amount * price;
    
}, initialValue);
```

Reply:
@akosasante I am trying your code, maybe I am doing smth wrong but it gives me an error 
![image](https://github.com/NatachaKey/ecommerce/assets/103755090/18b6d19e-ff31-4613-9803-9a9614bc6e2d)
so I've tried adding async and )  and I got
![image](https://github.com/NatachaKey/ecommerce/assets/103755090/e7e6c641-986f-4e73-90df-d02c5e5c5e39)
in postman  I get
![image](https://github.com/NatachaKey/ecommerce/assets/103755090/6d000668-0372-47bf-979c-fa01b69c6997)
what am I doing wrong?

Reply:
Oops no that's totally a mistake on my end. I talked about using `Array.reduce` but then I wrote `.map` 🤦

Reply:
I think I made multiple mistakes in my suggestion actually; like also forgetting to make sure that the function in the reduce is `async`. In the end; I think the for...of loop that you had is probably more readable/better fit for this use case. But still, included the reduce approach in this commit so you can take a look and compare for yourself: https://github.com/NatachaKey/ecommerce/pull/1/files#diff-15b17d867f3ec3b7bd415abe71693890cfcbff8c9099791377f97f9e90978e22R24-R58

Reply:
@akosasante thank you so much!! especially for consolelogging what we get in every step!!! that really helps to understand what's happenning